### ISSUE TITLE
Crawl never exits

### ISSUE THREAD

**Issue by rgaudin** [webrecorder/browsertrix-crawler#146]
Launched a run via zimit using 0.7.0.beta.1 and the crawl process never exited.

```
Running browsertrix-crawler crawl: crawl --newContext page --waitUntil load,networkidle0 --depth -1 --timeout 90 --behaviors autoplay,autofetch,siteSpecific --behaviorTimeout 90 --sizeLimit 4294967296 --timeLimit 7200 --url https://www.abc.com.py/ --userAgentSuffix +Zimit contact+zimfarm@kiwix.org --cwd /output/.tmpzbb8lh9v --statsFilename /output/crawl.json
```

```
Time threshold reached 7272.276 > 7200, stopping

== Start:     2022-07-10 12:12:49.180
== Now:       2022-07-10 14:13:59.737 (running for 2.0 hours)
== Progress:  204 / 204 (100.00%), errors: 0 (0.00%)
== Remaining: 0.0 ms (@ 0.03 pages/second)
== Sys. load: 21.8% CPU / 22.4% memory
== Workers:   1
   #0 IDLE
Saving crawl state to: /output/.tmpzbb8lh9v/collections/crawl-20220710121247300/crawls/crawl-20220710141359-7facd8c44598.yaml
Waiting to ensure pending data is written to WARCs...
```

It's been close to 24h now… It's not the first time I see this behavior. What could cause this to happen?

**Comment by ikreymer** [comment]
It looks like it got to the 'waiting to ensure pending data is written to WARCs...' so probably that check is stalling somehow... The above was the end of the output, right? The data is stored in redis, and when logging pywb, redis is also logged to `./logs/redis.log`.
Perhaps there is some info that can be gleaned from that?
Will try to take a look.

**Comment by rgaudin** [comment]
Yes that was the end of the output ; as for the other logs, I don't have access to those anymore. Note that I haven't seen it again since this report but it's not frequent anyway.

**Comment by wsdookadr** [comment]
I've thought about this and wrote [some notes](https://wsdookadr.github.io/posts/p8/) and a [small implementation](https://github.com/wsdookadr/femtocrawl) using only mitmproxy and a browser.

In mitmproxy for example, who also has to wait for pending requests, IPC signals [are handled](https://github.com/mitmproxy/mitmproxy/blob/156089229f76a5a4b08c0ffd492abad2a93e64ac/mitmproxy/tools/main.py#L109) and upon receiving a [shutdown event](https://github.com/mitmproxy/mitmproxy/blob/e11626395e42113f43cb6c94feab28c977389cd8/mitmproxy/master.py#L61) is triggered. This will release a [wait call](https://github.com/mitmproxy/mitmproxy/blob/e11626395e42113f43cb6c94feab28c977389cd8/mitmproxy/master.py#L54) here which in turn will continue the execution of the Done hooks, allowing whatever mitmproxy addons to write their data to disk.

Currently browsertrix-crawler [spawns puppeteer](https://github.com/webrecorder/browsertrix-crawler/blob/e22d95e2f07ef8a4cd3a4c309ee9ca0d6bab559e/crawler.js#L286) but tells it not to handle IPC signals. Maybe this is because [browsertrix wants to handle SIGINT](https://github.com/webrecorder/browsertrix-crawler/blob/cf90304fa736bef1443f35a943a1700491e1e8ad/main.js#L9) itself in a graceful manner. But this graceful manner means waiting for the pending requests and since it's waiting for those requests it won't respect the timeout.

I've a suggestion here. When SIGINT is received:
* browsertrix kills its puppeteer instances
* browsertrix cancells or marks as done all pending requests (because whatever data was in transit at that point was already intercepted by pywb)

This seems a lot like what @ikreymer is describing in webrecorder/browsertrix-cloud#298

**Comment by ikreymer** [comment]
@rgaudin I think this was likely fixed via 5c931275ed9077e43906be7eefb473c0b644ed2f, previously the crawler would wait indefinitely for pywb to finish writing WARC records, in some cases this could result it in never finishing..

@wsdookadr This is a slightly different issue, dealing with interrupts. Yes, currently the interrupt waits for the page to finish. Perhaps could add a separate mode for more abrupt interruption. The logic is a bit confusing there and could probably use more documentation..

I'll close this for now as I suspect its fixed, but feel free to comment if it happens again.
